### PR TITLE
Attentionbody cuda

### PR DESCRIPTION
**Pull request by almaudoh**
Finally got smolgen, input gating and sqr relu working on ankan's attention body branch PR #1713. BT2 runs with very minor differences from metal backend in policy, value and moves left

### CODE REVIEW

**Comment by borg323** [comment]
You need something like the following to fix the cudnn build:
```diff
--- a/src/neural/cuda/network_cudnn.cc
+++ b/src/neural/cuda/network_cudnn.cc
@@ -504,7 +504,7 @@ class CudnnNetwork : public Network {
     // Policy head.
     if (attn_policy_) {
       auto AttentionPolicy = std::make_unique<AttentionPolicyHead<DataType>>(
-          getLastLayer(), weights, scratch_mem_);
+          getLastLayer(), weights, scratch_mem_, false, SELU);
       network_.emplace_back(std::move(AttentionPolicy));
 
       auto policymap = std::make_unique<PolicyMapLayer<DataType>>(
```

**Comment by almaudoh** [comment]
> You need something like the following to fix the cudnn build:
> 
> ```diff
> --- a/src/neural/cuda/network_cudnn.cc
> +++ b/src/neural/cuda/network_cudnn.cc
> @@ -504,7 +504,7 @@ class CudnnNetwork : public Network {
>      // Policy head.
>      if (attn_policy_) {
>        auto AttentionPolicy = std::make_unique<AttentionPolicyHead<DataType>>(
> -          getLastLayer(), weights, scratch_mem_);
> +          getLastLayer(), weights, scratch_mem_, false, SELU);
>        network_.emplace_back(std::move(AttentionPolicy));
>  
>        auto policymap = std::make_unique<PolicyMapLayer<DataType>>(
> ```

Fixed.

File: src/neural/cuda/layers.cc [hunk 1]
```diff
@@ -1733,8 +1937,178 @@ AttentionPolicyHead<DataType>::EncoderWeights::~EncoderWeights() {
...
+
+template <typename DataType>
+void AttentionBody<DataType>::Eval(
+    int N, DataType* output, const DataType* input, const DataType* input2,
+    void* scratch, size_t scratch_size, cudnnHandle_t /*cudnn*/,
+    cublasHandle_t cublas, cudaStream_t stream) {
+  DataType* scratch0 = (DataType*)scratch;
+  DataType* scratch1 = (DataType*)output;
+  DataType* scratch2 = (DataType*)input2;
+  DataType* scratch3 = scratch2 + scratch_size / (2 * sizeof(DataType));
```
Comment (borg323) [inline src/neural/cuda/layers.cc:2058]: Does this have to be relative to input2 or it could have been relative to output as well?

**Review by borg323 — COMMENTED** [review]
There are several warnings about the initialization order of variables, and some lines need formatting.

File: src/neural/cuda/network_cuda.cc [hunk 2]
```diff
@@ -82,6 +83,39 @@ static size_t getMaxAttentionHeadSize(const LegacyWeights& weights, int N) {
   return size;
 }
 
+static size_t getMaxAttentionBodySize(const LegacyWeights& weights, int N) {
+  const size_t embedding_op_size = weights.ip_emb_b.size();
+
+  size_t encoder_d_model = 0;
+  size_t encoder_dff = 0;
```
[the remark this answers is not in the document]
Comment (almaudoh) [reply src/neural/cuda/network_cuda.cc:90]: Nice catch! That should be part of the max scratch size calculation.

File: src/neural/cuda/layers.cc [hunk 3]
```diff
@@ -1487,18 +1578,264 @@ static void cublasXGemmStridedBatched(
...
+        (float**)B, ldb, &beta, (float**)C, ldc,
+        batchCount));
+  }
+}
+
+// input/output tensor is scratch1, others are used as scratch.
+// TODO: fix naming of scratch buffers
+template <typename DataType>
+void EncoderBlock<DataType>::Eval(int N, DataType* scratch1, DataType* scratch0,
+                                  DataType* scratch2, DataType* scratch3,
```
[the remark this answers is not in the document]
Comment (almaudoh) [reply src/neural/cuda/layers.cc:1607]: I have renamed `scratch1` to reflect its function as the buffer for the input and output. The rest are reusable buffers and are used in different ways. Not sure how best to name them for easier understanding of the code.

**Pull request merged**